### PR TITLE
Fix validation order and major step towards fixing #979

### PR DESCRIPTION
**Pull request by kkohbrok**
This PR is a major step towards bringing OpenMLS up to speed with the draft-16 framing structure. The PR is large due to the way the validation chain works. Previously, an MLSPlaintext was the result of the chain, now it is one of the inputs and an MLSAuthenticatedContent is the result of the chain.

This PR partly reverses some changes from my previous PRs, since I now better understand what has to go where. That being said, this is not going to be the last PR in the series. (Even though it should be the last substantial one.) Since I want to do a follow-up PR for cleanup and to restore the abstraction boundaries (see comments below), I don't want to mark any issues as solved by this PR, but will instead do that manually after this PR is merged.

This PR should be squashed, the commits are only for review purposes. They are not super precise in encapsulating individual changes across the whole code base, but they contain the major changes followed by commits that contain mostly churn, so it is probably helpful to review commit-by-commit and in-order. Here are some notes regarding the changes in general. Please read before reviewing.

* The abstraction boundaries that enforce the validation chain are slightly broken. I intend to fix this in the next follow up PR.
  * For now, MLSPlaintext fields are pub(super) to facilitate TLS codec implementations. They should be made private.
* The validation chain makes more sense now and reflects the figure in section 7 of the spec. The MLSPlaintext is now on the "outside", where it was previously both an input to the validation chain and an output. Now it's only an input with MLSAuthenticatedContent (a newly introduced struct) being the output. We could have used MLSContent, but we need the data in MLSAuthenticatedContent for further validation.
* The MembershipTag is now checked when processing the MLSPlaintext, i.e. before signature verification, and added after turning an MLSAuthenticatedContent into an MLSPlaintext
  * It would be nice to turn membership tag computation and validation into a type-encoded thing similar to how we handle signatures, but I leave that to a follow-up. (Same for confirmation tag.)
* MLSPlaintext and VerifiableMlsAuthContent swapped places in a number of locations. Some large parts of the corresponding implementations also swapped between the two structs as a consequence.
* WireFormat does not work the way it worked before. Previously, it indicated if a plaintext was originally a plaintext or if it is the result of decrypting a ciphertext. Now it is not encrypted or otherwise sent over the wire. Instead it is included in the TBS and thus checked implicitly by validating the signature. This PR thus removes ValSem001. The test that previously tested for the correct WireFormat now checks if the signature fails if the WireFormat that was originally signed is different from the WireFormat used.
* I kept renaming to a minimum so the PR doesn't get too big. There are still a number of places where variables or functions are called "...plaintext..." when it's actually MLSAuthenticatedContent or MLSContent.
* I removed the explicit check for the presence of a confirmation tag, as it's no longer encoded as an optional field, but a select conditioned on the content type. As such a missing confirmation tag leads to a decoding error.

* I slightly changed how the encryption function resulting in an MLSCiphertext is called (but not the actual content). Previously it took a header as additional input, which was only needed for some tests. I refactored the code and introduced a function that allows manipulation of the header only for tests.
  * The changes lead to minor changes how a modified sender is passed to the encryption function.
* I tried to put all of what I thought was churn into one commit to single out the important changes.
  * I refactored some of the test (mostly de-duplicating some code) to better cope with the churn. For example, I introduced a re-signing function for a number of tests.
  * A lot of the churn is due to the membership tag being verified before most other checks. That means that it now has to be set explicitly after modifying a message for a test such that we can test the intended check.

### CODE REVIEW

**Comment by franziskuskiefer** [comment]
@kkohbrok can you resolve the conflicts and merge main? This is currently broken due to the tls_codec changes.

File: openmls/src/framing/errors.rs [hunk 1]
```diff
@@ -77,3 +77,14 @@ pub enum MlsMessageError {
     #[error("The message could not be decoded.")]
     UnableToDecode,
 }
+
+/// Error when fetching credential from non-member message
+#[derive(Error, Debug, PartialEq, Clone)]
+pub enum NonMemberCredentialError {
```
Comment (franziskuskiefer) [inline openmls/src/framing/errors.rs:83]: This appears to be unused.

File: openmls/src/framing/message.rs [hunk 2]
```diff
@@ -28,7 +28,7 @@ use crate::error::LibraryError;
 ///     // ... continued in [MlsMessageBody] ...
 /// } MLSMessage;
 /// ```
-#[derive(PartialEq, Debug, Clone)]
+#[derive(PartialEq, Debug, Clone, TlsSerialize, TlsSize, TlsDeserialize)]
 pub(crate) struct MlsMessage {
     pub(crate) body: MlsMessageBody,
```
Comment (franziskuskiefer) [inline openmls/src/framing/message.rs:33]: I presume the version is added later?

File: openmls/src/framing/validation.rs [hunk 3]
```diff
@@ -265,30 +268,16 @@ impl UnverifiedContextMessage {
     ///  - ValSem008
```
Comment (franziskuskiefer) [inline openmls/src/framing/validation.rs:268]: ValSem008 doesn't appear to be in here anymore.

File: openmls/src/group/core_group/process.rs [hunk 4]
```diff
@@ -34,14 +34,25 @@ impl CoreGroup {
...
-            WireFormat::MlsPlaintext => DecryptedMessage::from_inbound_plaintext(message)?,
-            WireFormat::MlsCiphertext => {
+        //  - ValSem007 MembershipTag presence
+        let decrypted_message = match message.mls_message.body {
+            MlsMessageBody::Plaintext(plaintext) => {
+                // If the message is older than the current epoch, we need to fetch the correct secret tree first.
+                let message_secrets =
+                    self.message_secrets_for_epoch(epoch).map_err(|e| match e {
+                        SecretTreeError::TooDistantInThePast => ValidationError::NoPastEpochData,
+                        _ => LibraryError::custom("Unexpected return value").into(),
```
Comment (franziskuskiefer) [inline openmls/src/group/core_group/process.rs:48]: Maybe have a message with a little more information about what went wrong? Something like "Unexpected error while retrieving message secrets for epoch".

File: openmls/src/group/core_group/create_commit.rs [hunk 5]
```diff
@@ -223,13 +223,13 @@ impl CoreGroup {
         let mut provisional_epoch = self.group_context.epoch();
         provisional_epoch.increment();
 
-        // Build MlsPlaintext
-        let mut mls_plaintext = MlsPlaintext::commit(
+        // Build MlsAuthContent
+        let mut content = MlsAuthContent::commit(
```
Comment (franziskuskiefer) [inline openmls/src/group/core_group/create_commit.rs:227]: Maybe just call it `commit` Makes creating the struct further down easier and is more descriptive.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, quite a lot of churn but nothing too complicated. It generally looks good to me.
I didn't check the tests yet. I'll look at them later.

Can you udpate the description in `validation.rs` to map the new flow if anything changed there? Or more general describe the message flow somewhere, similar to the figure in sec 7 in the spec but in the way we implement it? I think this will make it easier for anyone, including future self, to understand the flows.

File: openmls/src/framing/ciphertext.rs [hunk 6]
```diff
@@ -70,21 +65,91 @@ impl MlsCiphertext {
...
+            mls_plaintext,
+            ciphersuite,
+            backend,
+            message_secrets,
+            padding_size,
+        )
+    }
+
+    #[cfg(test)]
+    pub(crate) fn encrypt_without_check(
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/framing/ciphertext.rs:93]: Sure, I'll file an issue and link it here.

File: .github/workflows/interop_client.yml [hunk 7]
```diff
@@ -20,15 +20,15 @@ defaults:
...
-      fail-fast: false
-      matrix:
-        os:
-          - ubuntu-latest
-    runs-on: ${{ matrix.os }}
-    steps:
-      - uses: actions/checkout@v2
-      - name: Run interop client test script
-        run: ./test_with_runner.py
+#jobs:
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply .github/workflows/interop_client.yml:23]: I'll find a good issue to track this in and link it here.

File: openmls/src/framing/ciphertext.rs [hunk 8]
```diff
@@ -70,21 +65,91 @@ impl MlsCiphertext {
...
+            backend,
+            message_secrets,
+            padding_size,
+        )
+    }
+
+    /// Internal function to encrypt content. The extra message header is only used
+    /// for tests. Otherwise, the data from the given `MlsAuthContent` is used.
+    fn encrypt_content(
+        test_header: Option<MlsMessageHeader>,
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/framing/ciphertext.rs:132]: As I said in the other comment. I'll file a follow-up.

File: openmls/src/framing/plaintext.rs [hunk 9]
```diff
@@ -526,36 +461,213 @@ pub(crate) struct MlsContentAuthData {
     pub(super) confirmation_tag: Option<ConfirmationTag>,
 }
 
-#[cfg(test)]
-impl MlsContentAuthData {
-    pub fn new(signature: Signature, confirmation_tag: impl Into<Option<ConfirmationTag>>) -> Self {
+#[derive(PartialEq, Debug, Clone, TlsSerialize, TlsSize)]
+pub(crate) struct MlsAuthContent {
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/framing/plaintext.rs:465]: Maybe we can make the VerifiableMlsAuthContent a wrapper struct around this one? I'll give this some more thought and I'll come up with a proposal.

File: openmls/src/framing/ciphertext.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/framing/ciphertext.rs:93]: I filed #1148.

File: .github/workflows/interop_client.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply .github/workflows/interop_client.yml:23]: I filed ##1149.

File: openmls/src/framing/message.rs (hunk 2 above)
Comment (kkohbrok) [reply openmls/src/framing/message.rs:33]: Yes, it's probably a relatively small change, but I didn't want to stretch this PR further. It's tracked in #979 and I'll tackle it in a follow-up.

**Review by kkohbrok — COMMENTED** [review]
Thanks for the review! I think I addressed every comment. Let me know if there are more changes you want me to make.

**Review by franziskuskiefer — APPROVED** [review]
Thanks, let's get this in.

We should really add something like figure 7 from the spec with the OpenMLS types. But you can do this in one of the follow-ups.

File: openmls/src/framing/test_framing.rs [hunk 10]
```diff
@@ -57,19 +59,23 @@ fn codec_plaintext(ciphersuite: Ciphersuite, backend: &impl OpenMlsCryptoProvide
...
         .sign(backend, &credential_bundle)
-        .expect("Signing failed.");
+        .expect("Signing failed.")
+        .into();
+
+    let membership_key = MembershipKey::from_secret(
+        Secret::random(ciphersuite, backend, None /* MLS version */)
+            .expect("Not enough randomness."),
+    );
+    orig.set_membership_tag(backend, &serialized_context, &membership_key)
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/framing/test_framing.rs:71]: I agree that it should be enforced on a type level, which is tracked in #430. I'd say we leave that out of #979 for the time being.

**Review by franziskuskiefer — APPROVED** [review]

**Pull request merged**